### ISSUE TITLE
[Errno 97] Address family not supported by protocol

### ISSUE THREAD

**Issue by ongsunglau** [NagiosEnterprises/ncpa#708]
Hi There,

Recently I found out NCPA 2.2.2 have the issue below, it seems like Python issue but when I checked below thread is because of the NCPA pointing to IPv6 by default. I'm not sure this happen to others or not but current server which installed ncpa is Physical server. So do I need to update the python or not suppose to change that default IP pointing?
https://support.nagios.com/forum/viewtopic.php?f=7&t=56125 

ncpa.cfg
```
#
#   NCPA Main Config File
#   ---------------------
#

#
# -------------------------------
# General Configuration
# -------------------------------
#

[general]

#
# Check logging (in ncpa.db and the interface) is on by default, you can disable it
# if you do not want to record the check requests that are coming in or checks being
# sent over NRDP.
# Default: check_logging = 1
#
check_logging = 1

#
# Check logging time - how long in DAYS you'd like to keep checks in the database.
# Default: 30
#
check_logging_time = 30

#
# Display all mounted disk partitions
# (essentially setting all=True here: https://psutil.readthedocs.io/en/latest/#psutil.disk_partitions)
# Default: 1
#
all_partitions = 1

#
# Excluded file system types removes these fs types from the disk metrics
# (This is mostly only noteable on UNIX systems but also works on Windows if you need it)
# Default: aufs,autofs,binfmt_misc,cifs,cgroup,configfs,debugfs,devpts,devtmpfs,
#          encryptfs,efivarfs,fuse,fusectl,hugetlbfs,mqueue,nfs,overlayfs,proc,pstore,
#          rpc_pipefs,securityfs,selinuxfs,smb,sysfs,tmpfs,tracefs
#
exclude_fs_types = aufs,autofs,binfmt_misc,cifs,cgroup,configfs,debugfs,devpts,devtmpfs,encryptfs,efivarfs,fuse,fusectl,hugetlbfs,mqueue,nfs,overlayfs,proc,pstore,rpc_pipefs,securityfs,selinuxfs,smb,sysfs,tmpfs,tracefs

#
# The default unit to convert bytes (B) into if no unit is specified
# (Gi = 1024 MiB, G = 1000 MB)
#
default_units = Gi

#
# -------------------------------
# Listener Configuration (daemon)
# -------------------------------
#

[listener]

#
# User and group to run plugins as (recommended to use nagios:nagios)
# Default: uid = nagios
# Default: gid = nagios
#
# ** Note - The daemon runs as root, but forks a child process when running a plugin
#    that is defined by the user, for security reasons. However, without the main daemon
#    running as root, much of the system information would be missing. This is typical behavior. **
#
# This is for Unix only (Linux, Mac OS X, etc)
#
uid = nagios
gid = nagios

#
# IP address and port number for the Listener to use for the web GUI and API
#
# :: allows for dual stack (IPv4 and IPv6 on most linux systems) but will only allow
# for IPv6 connections on Windows
# 0.0.0.0 allows for IPv4 connections only on Windows and most linux systems
#
# Default: ip = ::
# Default (Windows): ip = 0.0.0.0
# Default: port = 5693
#
# ip =
# port =
**ip = 0.0.0.0** <-- After put this line only can works
#
# SSL connection and certificate config (if an SSL option is not available on some older
# operating systems it will default back to TLSv1)
# ssl_version options: TLSv1, TLSv1_1, TLSv1_2
#
# ssl_ciphers = <list of ciphers>
#
ssl_version = TLSv1_2
certificate = adhoc

#
# Listener logging file level, location, and the PID location
# Default: loglevel = info (debug, info, warning, error)
# Default: logfile = var/log/ncpa_listener.log
# Default: pidfile = var/run/ncpa_listener.pid (leave listener in pid file name)
#
loglevel = info
logfile = var/log/ncpa_listener.log
pidfile = var/run/ncpa_listener.pid

#
# Delay the listener (API & web GUI) from starting in seconds
# Default: 0
#
# delay_start = 30

#
# Allow admin functionality in the web GUI. When this is set to 0, the admin section will not
# be displayed in the header and will not be available to be accessed.
# Default: 1
#
admin_gui_access = 1

#
# Admin password for the admin section in the web GUI, by default there is no admin
# password and the admin section of the GUI can be accessed by anyone if admin_gui_access is set to 1.
# Default: None
#
# Note: Setting this value to 'None' will automatically log you in, setting it empty will allow you to
# log in using a blank password.
#
admin_password = None

#
# Require admin password to access ALL of the web GUI.
# This does not affect API access via token (community_string).
# Default: 0
#
admin_auth_only = 0

#
# Comma separated list of allowed hosts that can access the API (and GUI)
# Exmaple: 192.168.23.15
# Example subnet: 192.168.0.0/28
#
# allowed_hosts =

#
# Number of maximum concurrent connections to the NCPA server.
# Use "None" for unlimited. Default is 200.
# Example: 200
#
# max_connections =

#
# Set the URL to use in the X-Frame-Options and Content-Security-Policy headers
# in order to enable the NCPA GUI to be allowed to load intp a frame
# Default: None
# Example: mycoolwebsite.com
# Example: *.mycoolwebsite.com
#
# allowed_sources =

#
# -------------------------------
# Listener Configuration (API)
# -------------------------------
#

[api]

#
# The token that will be used to log into the basic web GUI (API browser, graphs, top charts, etc)
# and to authenticate requests to the API and requests through check_ncpa.py
#
community_string = nagiosadmin

#
# -------------------------------
# Passive Configuration (daemon)
# -------------------------------
#

[passive]

#
# Handlers are a comma separated list of what you would like the passive agent to run
# Default: None
# Options:
#   nrds, nrdp, kafkaproducer
#
# Example:
# handlers = nrds,nrdp,kafkaproducer
#
handlers = None

#
# User and group to run passive checks as (Recommended to use nagios:nagios)
# Default: uid = nagios
# Default: gid = nagios
#
uid = nagios
gid = nagios

#
# Passive check interval - the amount in seconds to wait between each passive check by default,
# this can be overwritten by adding on a "|<duration>" in seconds to the passive check config
# Default: 300 (5 minutes)
#
sleep = 300

#
# Passive logging file level, location, and the PID location
# Default: loglevel = info (debug, info, warning, error)
# Default: logfile = var/log/ncpa_passive.log
# Default: pidfile = var/run/ncpa_passive.pid (leave passive in pid file name)
#
loglevel = info
logfile = var/log/ncpa_passive.log
pidfile = var/run/ncpa_passive.pid

#
# Delay passive checks from starting in seconds
# Default: 0
#
# delay_start = 30

#
# -------------------------------
# Passive Configuration (NRDP)
# -------------------------------
#

[nrdp]

#
# Connection settings to the NRDP server
# parent = NRDP server location (ex: http://<address>/nrdp)
# token = NRDP server token used to send NRDP results
#
parent =
token =

#
# The hostname that will replace %HOSTNAME% in the check definitions and will be
# sent to NRDP with the check name as the service description (service name)
#
hostname = NCPA 2

#
# -------------------------------
# Passive Configuration (NRDS)
# -------------------------------
#

[nrds]

#
# NRDS CONFIGURATION DOES NOT WORK YET. MORE TO COME IN VERSION 2.1.0.
#

#
# NRDS connection information
#
url =
token =
config_name =
config_version =
update_config = 1
update_plugins = 1

[kafkaproducer]

#
# -------------------------------
# Passive Configuration (Kafka)
# -------------------------------
#

hostname = None
servers = localhost:9092
clientname = NCPA-Kafka
topic = ncpa

#
# -------------------------------
# Plugin Configuration
# -------------------------------
#

[plugin directives]

#
# Plugin path where all plugins will be ran from.
#
plugin_path = plugins/

#
# Plugin execution timeout in seconds. Different than the check_ncpa.py timeout, which is
# normally for network connection issues. Will return a CRITICAL value and error when the plugin
# reaches the defined max execution timeout and kills the process.
# Default: 60
#
# plugin_timeout = 60

#
# Extensions for plugins
# ----------------------
# The extension for the plugin denotes how NCPA will try to run the plugin. Use this
# for setting how you want to run the plugin in the command line.
#
# NOTE: Plugins without an extension will be ran in the cmdline as follows:
#       $plugin_name $plugin_args
#
# Defaults:
# .sh = /bin/sh $plugin_name $plugin_args
# .py = python $plugin_name $plugin_args
# .ps1 = powershell -ExecutionPolicy Bypass -File $plugin_name $plugin_args
# .vbs = cscript $plugin_name $plugin_args //NoLogo
# .bat = cmd /c $plugin_name $plugin_args
#
# Since windows NCPA is 32-bit, if you need to use 64-bit powershell, try the following for
# the powershell plugin definition:
# .ps1 = c:\windows\sysnative\windowspowershell\v1.0\powershell.exe -ExecutionPolicy Unrestricted -File $plugin_name $plugin_args
#

# Linux / Mac OS X
.sh = /bin/sh $plugin_name $plugin_args
.py = python $plugin_name $plugin_args

# Windows
.ps1 = powershell -ExecutionPolicy Bypass -File $plugin_name $plugin_args
.vbs = cscript $plugin_name $plugin_args //NoLogo
.wsf = cscript $plugin_name $plugin_args //NoLogo
.bat = cmd /c $plugin_name $plugin_args
```

```
2020-11-02 15:31:42,399 27018 INFO stopped
2020-11-02 15:38:50,554 28399 INFO started
2020-11-02 15:38:50,555 28399 INFO Using SSL version TLSv1_2
2020-11-02 15:38:50,557 28399 ERROR [Errno 97] Address family not supported by protocol
Traceback (most recent call last):
  File "ncpa_listener.py", line 108, in run
  File "/usr/local/lib/python2.7/site-packages/gevent/baseserver.py", line 360, in serve_forever
  File "/usr/local/lib/python2.7/site-packages/gevent/baseserver.py", line 304, in start
  File "/usr/local/lib/python2.7/site-packages/gevent/pywsgi.py", line 1450, in init_socket
  File "/usr/local/lib/python2.7/site-packages/gevent/server.py", line 127, in init_socket
  File "/usr/local/lib/python2.7/site-packages/gevent/server.py", line 138, in get_listener
  File "/usr/local/lib/python2.7/site-packages/gevent/server.py", line 225, in _tcp_listener
  File "/usr/local/lib/python2.7/site-packages/gevent/_socket2.py", line 107, in __init__
error: [Errno 97] Address family not supported by protocol
2020-11-02 15:38:50,557 28399 INFO stopped
2020-11-02 15:42:23,445 29213 INFO started
2020-11-02 15:42:23,446 29213 INFO Using SSL version TLSv1_2
2020-11-02 15:42:23,448 29213 ERROR [Errno 97] Address family not supported by protocol
Traceback (most recent call last):
  File "ncpa_listener.py", line 108, in run
  File "/usr/local/lib/python2.7/site-packages/gevent/baseserver.py", line 360, in serve_forever
  File "/usr/local/lib/python2.7/site-packages/gevent/baseserver.py", line 304, in start
  File "/usr/local/lib/python2.7/site-packages/gevent/pywsgi.py", line 1450, in init_socket
  File "/usr/local/lib/python2.7/site-packages/gevent/server.py", line 127, in init_socket
  File "/usr/local/lib/python2.7/site-packages/gevent/server.py", line 138, in get_listener
  File "/usr/local/lib/python2.7/site-packages/gevent/server.py", line 225, in _tcp_listener
  File "/usr/local/lib/python2.7/site-packages/gevent/_socket2.py", line 107, in __init__
error: [Errno 97] Address family not supported by protocol
```

```
[root@server  ~]# systemctl status ncpa_listener
● ncpa_listener.service - LSB: This manages the NCPA Listener service
   Loaded: loaded (/etc/rc.d/init.d/ncpa_listener; bad; vendor preset: disabled)
   Active: active (exited) since Mon 2020-11-02 15:42:23 +08; 11s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 29119 ExecStop=/etc/rc.d/init.d/ncpa_listener stop (code=exited, status=0/SUCCESS)
  Process: 29200 ExecStart=/etc/rc.d/init.d/ncpa_listener start (code=exited, status=0/SUCCESS)

Nov 02 15:42:23 server systemd[1]: Starting LSB: This manages the NCPA Listener service...
Nov 02 15:42:23 server  ncpa_listener[29200]: Starting NCPA Listener: [  OK  ]
Nov 02 15:42:23 server  systemd[1]: Started LSB: This manages the NCPA Listener service.
```

```
[root@server log]# rpm -qa |grep -i python-2.7
python-2.7.5-88.el7.x86_64
[root@server log]# rpm -qa| grep ncpa
ncpa-2.2.2-1.el7.x86_64
```

**Comment by jomann09** [comment]
NCPA uses an internal version of Python so changing the version won't affect it. By default if there is no IP address specified with the  `ip` variable it will use `::` and not `0.0.0.0` because on Linux `::` is supposed to bind to both IPv4 and IPv6. I don't actually know why your system would not bind to both, my networking knowledge is pretty basic.

**Comment by ongsunglau** [comment]
Is that could be Physical server? Because this not happen in VM host but only newly setup Physical server.

**Comment by jomann09** [comment]
I have tried it on a physical server and didn't have the issue, so I'm just thinking it's related to how the network on the particular server was set up.

**Comment by ongsunglau** [comment]
Well, initially this is RHEL server, so we setup using NetworkManager #nmtui. I'm not sure is that related or not. Because we will configure IP address using nmtui and restart it.

**Comment by ccztux** [comment]
This should be fixed in version 2.3.0 https://github.com/NagiosEnterprises/ncpa/commit/f079521d0a57d984c2caec88375286238a435f64.
Can you please test it and give us feedback?

**Comment by jomann09** [comment]
Closing this due to the fix in 2.3, if you're still having problems feel free to re-open this issue or start a new one.
